File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/json_stream_parser.ts
/**
 * A streaming JSON parser optimized for LLM responses.
 *
 * This parser processes JSON data character-by-character as it streams in,
 * allowing you to react to properties before the entire JSON is received.
 * It's specifically designed for handling Large Language Model (LLM) streaming
 * responses that output structured JSON data.
 */

import {
    ArrayPropertyStreamController,
    BooleanPropertyStreamController,
    NullPropertyStreamController,
    NumberPropertyStreamController,
    ObjectPropertyStreamController,
    PropertyStreamController,
    StringPropertyStreamController,
} from "./property_stream_controller.js";
import {
    ArrayPropertyStream,
    BooleanPropertyStream,
    NullPropertyStream,
    NumberPropertyStream,
    ObjectPropertyStream,
    PropertyStream,
    StringPropertyStream,
} from "./property_stream.js";
import { PropertyDelegate } from "./property_delegates/property_delegate.js";
import { ObjectPropertyDelegate } from "./property_delegates/object_property_delegate.js";
import { ArrayPropertyDelegate } from "./property_delegates/array_property_delegate.js";

/**
 * Controller interface for coordinating parsing operations.
 * Internal use only - not exposed in pub   lic API.
 */
export class JsonStreamParserController {
    constructor(
        private addPropertyChunkFn: <T>(params: {
            chunk: T;
            propertyPath: string;
        }) => void,
        private getPropertyStreamControllerFn: (
            propertyPath: string,
        ) => PropertyStreamController<any> | undefined,
        private getPropertyStreamFn: (
            propertyPath: string,
            streamType:
                | "string"
                | "number"
                | "boolean"
                | "null"
                | "object"
                | "array",
        ) => PropertyStream<any>,
        private completePropertyFn: <T>(propertyPath: string, value: T) => void,
    ) {}

    addPropertyChunk<T>(params: { chunk: T; propertyPath: string }): void {
        this.addPropertyChunkFn(params);
    }

    getPropertyStreamController(
        propertyPath: string,
    ): PropertyStreamController<any> | undefined {
        return this.getPropertyStreamControllerFn(propertyPath);
    }

    getPropertyStream(
        propertyPath: string,
        streamType:
            | "string"
            | "number"
            | "boolean"
            | "null"
            | "object"
            | "array",
    ): PropertyStream<any> {
        return this.getPropertyStreamFn(propertyPath, streamType);
    }

    completeProperty<T>(propertyPath: string, value: T): void {
        this.completePropertyFn(propertyPath, value);
    }
}

/**
 * Main streaming JSON parser class.
 */
export class JsonStreamParser {
    private controller!: JsonStreamParserController;
    private streamAbortController: AbortController | null = null;
    private propertyControllers: Map<string, PropertyStreamController<any>> =
        new Map();
    private disposed = false;
    private rootDelegate: PropertyDelegate | null = null;
    private closeOnRootComplete: boolean;
    private consumeStreamPromise: Promise<void> | null = null;

    constructor(
        stream: AsyncIterable<string>,
        options?: { closeOnRootComplete?: boolean },
    ) {
        this.closeOnRootComplete = options?.closeOnRootComplete ?? true;

        this.controller = new JsonStreamParserController(
            this.addPropertyChunk.bind(this),
            this.getControllerForPath.bind(this),
            this.getPropertyStreamForPath.bind(this),
            this.completePropertyAtPath.bind(this),
        );

        // Create an abort controller to stop the stream consumption
        this.streamAbortController = new AbortController();

        // Start consuming the stream
        this.consumeStreamPromise = this.consumeStream(stream);
    }

    private async consumeStream(stream: AsyncIterable<string>): Promise<void> {
        try {
            for await (const chunk of stream) {
                if (
                    this.disposed || this.streamAbortController?.signal.aborted
                ) {
                    break;
                }

                this.parseChunk(chunk);

                // Check if we should stop consuming after parsing
                if (
                    this.closeOnRootComplete &&
                    this.rootDelegate &&
                    this.rootDelegate.done
                ) {
                    this.streamAbortController?.abort();
                    break;
                }
            }

            // Stream ended normally
            this.handleStreamEnd();
        } catch (error) {
            // Stream error - reject all pending controllers
            for (const controller of this.propertyControllers.values()) {
                controller.completeError(
                    error instanceof Error ? error : new Error(String(error)),
                );
            }
        }
    }

    /**
     * Gets a stream for a string property at the specified propertyPath.
     */
    getStringProperty(propertyPath: string): StringPropertyStream {
        this.checkDisposed();
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            if (!(existing instanceof StringPropertyStreamController)) {
                throw new Error(
                    `Property at path ${propertyPath} is not a StringPropertyStream`,
                );
            }
            return existing.propertyStream;
        }

        const controller = new StringPropertyStreamController(
            this.controller,
            propertyPath,
        );
        this.propertyControllers.set(propertyPath, controller);
        return controller.propertyStream;
    }

    /**
     * Gets a stream for a number property at the specified propertyPath.
     */
    getNumberProperty(propertyPath: string): NumberPropertyStream {
        this.checkDisposed();
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            if (!(existing instanceof NumberPropertyStreamController)) {
                throw new Error(
                    `Property at path ${propertyPath} is not a NumberPropertyStream`,
                );
            }
            return existing.propertyStream;
        }

        const controller = new NumberPropertyStreamController(
            this.controller,
            propertyPath,
        );
        this.propertyControllers.set(propertyPath, controller);
        return controller.propertyStream;
    }

    /**
     * Gets a stream for a boolean property at the specified propertyPath.
     */
    getBooleanProperty(propertyPath: string): BooleanPropertyStream {
        this.checkDisposed();
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            if (!(existing instanceof BooleanPropertyStreamController)) {
                throw new Error(
                    `Property at path ${propertyPath} is not a BooleanPropertyStream`,
                );
            }
            return existing.propertyStream;
        }

        const controller = new BooleanPropertyStreamController(
            this.controller,
            propertyPath,
        );
        this.propertyControllers.set(propertyPath, controller);
        return controller.propertyStream;
    }

    /**
     * Gets a stream for a null property at the specified propertyPath.
     */
    getNullProperty(propertyPath: string): NullPropertyStream {
        this.checkDisposed();
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            if (!(existing instanceof NullPropertyStreamController)) {
                throw new Error(
                    `Property at path ${propertyPath} is not a NullPropertyStream`,
                );
            }
            return existing.propertyStream;
        }

        const controller = new NullPropertyStreamController(
            this.controller,
            propertyPath,
        );
        this.propertyControllers.set(propertyPath, controller);
        return controller.propertyStream;
    }

    /**
     * Gets a stream for an object property at the specified propertyPath.
     */
    getObjectProperty(propertyPath: string): ObjectPropertyStream {
        this.checkDisposed();
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            if (!(existing instanceof ObjectPropertyStreamController)) {
                throw new Error(
                    `Property at path ${propertyPath} is not a ObjectPropertyStream`,
                );
            }
            return existing.propertyStream;
        }

        const controller = new ObjectPropertyStreamController(
            this.controller,
            propertyPath,
        );
        this.propertyControllers.set(propertyPath, controller);
        return controller.propertyStream;
    }

    /**
     * Gets a stream for an array property at the specified propertyPath.
     */
    getArrayProperty<T = any>(propertyPath: string): ArrayPropertyStream<T> {
        this.checkDisposed();
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            if (!(existing instanceof ArrayPropertyStreamController)) {
                throw new Error(
                    `Property at path ${propertyPath} is not a ArrayPropertyStream`,
                );
            }
            return existing.propertyStream as ArrayPropertyStream<T>;
        }

        const controller = new ArrayPropertyStreamController<T>(
            this.controller,
            propertyPath,
        );
        this.propertyControllers.set(propertyPath, controller);
        return controller.propertyStream;
    }

    /**
     * Disposes the parser and cleans up resources.
     */
    async dispose(): Promise<void> {
        if (this.disposed) {
            return;
        }

        this.disposed = true;

        // Abort the stream consumption
        if (this.streamAbortController) {
            this.streamAbortController.abort();
        }

        // Wait for stream consumption to complete
        if (this.consumeStreamPromise) {
            try {
                await this.consumeStreamPromise;
            } catch {
                // Ignore errors during cleanup
            }
        }

        // Close all property controllers
        for (const controller of this.propertyControllers.values()) {
            if (!controller.isClosed) {
                controller.completeError(new Error("Parser disposed"));
            }
        }

        this.propertyControllers.clear();
    }

    private checkDisposed(): void {
        if (this.disposed) {
            throw new Error("Parser has been disposed");
        }
    }

    private parseChunk(chunk: string): void {
        if (this.disposed) return;

        // Check yap filter
        if (
            this.closeOnRootComplete &&
            this.rootDelegate &&
            this.rootDelegate.done
        ) {
            this.streamAbortController?.abort();
            return;
        }

        try {
            for (const character of chunk) {
                // Check yap filter inside loop
                if (
                    this.closeOnRootComplete &&
                    this.rootDelegate &&
                    this.rootDelegate.done
                ) {
                    this.streamAbortController?.abort();
                    return;
                }

                if (this.rootDelegate !== null) {
                    this.rootDelegate.addCharacter(character);
                    continue;
                }

                // Skip leading whitespace before root element
                if (/\s/.test(character)) {
                    continue;
                }

                if (character === "{") {
                    this.rootDelegate = new ObjectPropertyDelegate(
                        "",
                        this.controller,
                    );
                    this.rootDelegate.addCharacter(character);
                } else if (character === "[") {
                    this.rootDelegate = new ArrayPropertyDelegate(
                        "",
                        this.controller,
                    );
                    this.rootDelegate.addCharacter(character);
                }
            }

            // Notify delegates about chunk end
            this.rootDelegate?.onChunkEnd();

            // Final check after chunk
            if (
                this.closeOnRootComplete &&
                this.rootDelegate &&
                this.rootDelegate.done
            ) {
                this.streamAbortController?.abort();
            }
        } catch (e) {
            // Parsing error - already handled by completing specific controller with error
        }
    }

    private handleStreamEnd(): void {
        // Complete any incomplete property controllers with errors
        for (const controller of this.propertyControllers.values()) {
            if (!controller.isClosed) {
                controller.completeError(
                    new Error("Stream ended before property completed"),
                );
            }
        }
    }

    private addPropertyChunk<T>(params: {
        chunk: T;
        propertyPath: string;
    }): void {
        const controller = this.propertyControllers.get(params.propertyPath);
        if (!controller) return;

        if (controller instanceof StringPropertyStreamController) {
            controller.addChunk(params.chunk as string);
        } else if (controller instanceof ObjectPropertyStreamController) {
            // Objects emit snapshots - push to async iterator without completing
            controller.propertyStream._pushValue(
                params.chunk as Record<string, any>,
            );
        } else if (controller instanceof ArrayPropertyStreamController) {
            // Arrays emit snapshots - push to async iterator without completing
            controller.propertyStream._pushValue(params.chunk as any[]);
        }
        // Numbers, booleans, and nulls don't receive chunks - they complete directly
    }

    private getControllerForPath(
        propertyPath: string,
    ): PropertyStreamController<any> | undefined {
        return this.propertyControllers.get(propertyPath);
    }

    private getPropertyStreamForPath(
        propertyPath: string,
        streamType:
            | "string"
            | "number"
            | "boolean"
            | "null"
            | "object"
            | "array",
    ): PropertyStream<any> {
        // Check for existing controller with incompatible type
        const existing = this.propertyControllers.get(propertyPath);
        if (existing) {
            // Verify type compatibility
            const isCompatible = (streamType === "string" &&
                existing instanceof StringPropertyStreamController) ||
                (streamType === "number" &&
                    existing instanceof NumberPropertyStreamController) ||
                (streamType === "boolean" &&
                    existing instanceof BooleanPropertyStreamController) ||
                (streamType === "null" &&
                    existing instanceof NullPropertyStreamController) ||
                (streamType === "object" &&
                    existing instanceof ObjectPropertyStreamController) ||
                (streamType === "array" &&
                    existing instanceof ArrayPropertyStreamController);

            if (isCompatible) {
                return existing.propertyStream;
            } else {
                // Type mismatch - complete existing controller with error
                const error = new Error(
                    `Type mismatch at path "${propertyPath}": requested ${streamType} but found different type`,
                );
                existing.completeError(error);
                throw error;
            }
        }

        // Create appropriate stream based on type
        switch (streamType) {
            case "string":
                return this.getStringProperty(propertyPath);
            case "number":
                return this.getNumberProperty(propertyPath);
            case "boolean":
                return this.getBooleanProperty(propertyPath);
            case "null":
                return this.getNullProperty(propertyPath);
            case "object":
                return this.getObjectProperty(propertyPath);
            case "array":
                return this.getArrayProperty(propertyPath);
            default:
                throw new Error(`Unknown stream type: ${streamType}`);
        }
    }

    private completePropertyAtPath<T>(propertyPath: string, value: T): void {
        const controller = this.propertyControllers.get(propertyPath);
        if (!controller || controller.isClosed) return;

        if (controller instanceof StringPropertyStreamController) {
            controller.complete(value as unknown as string);
        } else if (controller instanceof NumberPropertyStreamController) {
            controller.complete(value as unknown as number);
        } else if (controller instanceof BooleanPropertyStreamController) {
            controller.complete(value as unknown as boolean);
        } else if (controller instanceof NullPropertyStreamController) {
            controller.complete(value as unknown as null);
        } else if (controller instanceof ObjectPropertyStreamController) {
            controller.complete(value as unknown as Record<string, any>);
        } else if (controller instanceof ArrayPropertyStreamController) {
            controller.complete(value as unknown as any[]);
        }
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/json_stream_parser.py
"""Streaming JSON parser optimized for LLM responses."""

from __future__ import annotations

import asyncio
from typing import Any, AsyncIterable, Callable, Dict, Optional, Tuple

from .parse_event import ParseEvent, ParseEventType
from .property_delegate import PropertyDelegate
from .property_delegates import ArrayPropertyDelegate, ObjectPropertyDelegate
from .property_stream import (
    PropertyGetterMixin,
    PropertyStream,
    StringPropertyStream,
    NumberPropertyStream,
    BooleanPropertyStream,
    NullPropertyStream,
    MapPropertyStream,
    ListPropertyStream,
)
from .property_stream_controller import (
    PropertyStreamController,
    StringPropertyStreamController,
    NumberPropertyStreamController,
    BooleanPropertyStreamController,
    NullPropertyStreamController,
    MapPropertyStreamController,
    ListPropertyStreamController,
)


class JsonStreamParserController:
    def __init__(
        self,
        add_property_chunk: Callable[[str, Any], None],
        get_property_stream_controller: Callable[
            [str], Optional[PropertyStreamController[Any]]
        ],
        get_property_stream: Callable[[str, str], PropertyStream[Any]],
        complete_property: Callable[[str, Any], None],
        emit_log: Callable[[ParseEvent], None],
    ) -> None:
        self._add_property_chunk = add_property_chunk
        self._get_property_stream_controller = get_property_stream_controller
        self._get_property_stream = get_property_stream
        self._complete_property = complete_property
        self._emit_log = emit_log

    def add_property_chunk(self, property_path: str, chunk: Any) -> None:
        self._add_property_chunk(property_path, chunk)

    def get_property_stream_controller(
        self, property_path: str
    ) -> Optional[PropertyStreamController[Any]]:
        return self._get_property_stream_controller(property_path)

    def get_property_stream(self, property_path: str, stream_type: str) -> PropertyStream[Any]:
        return self._get_property_stream(property_path, stream_type)

    def complete_property(self, property_path: str, value: Any) -> None:
        self._complete_property(property_path, value)

    def emit_log(self, event: ParseEvent) -> None:
        self._emit_log(event)


class JsonStreamParser(PropertyGetterMixin):
    """Main streaming JSON parser."""

    def __init__(
        self,
        stream: AsyncIterable[str],
        *,
        close_on_root_complete: bool = True,
        skip_thoughts: bool = False,
        thinking_tags: Tuple[str, str] = ("<think>", "</think>"),
        on_log: Optional[Callable[[ParseEvent], None]] = None,
    ) -> None:
        self.close_on_root_complete = close_on_root_complete
        self.skip_thoughts = skip_thoughts
        self.thinking_tags = thinking_tags
        self._on_log = on_log
        self._stream = stream
        self._property_controllers: Dict[str, PropertyStreamController[Any]] = {}
        self._root_delegate: Optional[PropertyDelegate] = None
        self._disposed = False
        self._abort_stream = False

        self._inside_thinking_tags = False
        self._tag_buffer = ""
        self._potential_tag_buffer = ""
        self._saw_potential_thinking_tags = False

        self._controller = JsonStreamParserController(
            self._add_property_chunk,
            self._get_controller_for_path,
            self._get_property_stream,
            self._complete_property,
            self._emit_log,
        )

        self._consume_task: asyncio.Task[None] = asyncio.create_task(
            self._consume_stream()
        )

    @property
    def parser_controller(self) -> JsonStreamParserController:
        return self._controller

    def build_property_path(self, key: str) -> str:
        return key

    async def _consume_stream(self) -> None:
        try:
            async for chunk in self._stream:
                if self._disposed or self._abort_stream:
                    break
                self._parse_chunk(chunk)

                if (
                    self.close_on_root_complete
                    and self._root_delegate
                    and self._root_delegate.done
                ):
                    self._abort_stream = True
                    break

            self._handle_stream_end()
        except Exception as exc:  # pragma: no cover - defensive
            for controller in self._property_controllers.values():
                controller.complete_error(
                    exc if isinstance(exc, Exception) else Exception(str(exc))
                )

    def get_string_property(self, property_path: str) -> StringPropertyStream:
        self._ensure_not_disposed()
        existing = self._property_controllers.get(property_path)
        if existing and not isinstance(existing, StringPropertyStreamController):
            raise ValueError(
                f"Property at path {property_path} is not a StringPropertyStream"
            )
        controller = self._property_controllers.setdefault(
            property_path,
            StringPropertyStreamController(self._controller, property_path),
        )
        return controller.property_stream  # type: ignore[return-value]

    def get_number_property(self, property_path: str) -> NumberPropertyStream:
        self._ensure_not_disposed()
        existing = self._property_controllers.get(property_path)
        if existing and not isinstance(existing, NumberPropertyStreamController):
            raise ValueError(
                f"Property at path {property_path} is not a NumberPropertyStream"
            )
        controller = self._property_controllers.setdefault(
            property_path,
            NumberPropertyStreamController(self._controller, property_path),
        )
        return controller.property_stream  # type: ignore[return-value]

    def get_boolean_property(self, property_path: str) -> BooleanPropertyStream:
        self._ensure_not_disposed()
        existing = self._property_controllers.get(property_path)
        if existing and not isinstance(existing, BooleanPropertyStreamController):
            raise ValueError(
                f"Property at path {property_path} is not a BooleanPropertyStream"
            )
        controller = self._property_controllers.setdefault(
            property_path,
            BooleanPropertyStreamController(self._controller, property_path),
        )
        return controller.property_stream  # type: ignore[return-value]

    def get_null_property(self, property_path: str) -> NullPropertyStream:
        self._ensure_not_disposed()
        existing = self._property_controllers.get(property_path)
        if existing and not isinstance(existing, NullPropertyStreamController):
            raise ValueError(
                f"Property at path {property_path} is not a NullPropertyStream"
            )
        controller = self._property_controllers.setdefault(
            property_path,
            NullPropertyStreamController(self._controller, property_path),
        )
        return controller.property_stream  # type: ignore[return-value]

    def get_map_property(self, property_path: str) -> MapPropertyStream:
        self._ensure_not_disposed()
        existing = self._property_controllers.get(property_path)
        if existing and not isinstance(existing, MapPropertyStreamController):
            raise ValueError(
                f"Property at path {property_path} is not a MapPropertyStream"
            )
        controller = self._property_controllers.setdefault(
            property_path,
            MapPropertyStreamController(self._controller, property_path),
        )
        return controller.property_stream  # type: ignore[return-value]

    def get_list_property(self, property_path: str) -> ListPropertyStream:
        self._ensure_not_disposed()
        existing = self._property_controllers.get(property_path)
        if existing and not isinstance(existing, ListPropertyStreamController):
            raise ValueError(
                f"Property at path {property_path} is not a ListPropertyStream"
            )
        controller = self._property_controllers.setdefault(
            property_path,
            ListPropertyStreamController(self._controller, property_path),
        )
        return controller.property_stream  # type: ignore[return-value]

    async def dispose(self) -> None:
        if self._disposed:
            return
        self._disposed = True
        self._abort_stream = True
        if not self._consume_task.done():
            self._consume_task.cancel()
            try:
                await self._consume_task
            except asyncio.CancelledError:
                pass
        for controller in self._property_controllers.values():
            if not controller.is_closed:
                controller.complete_error(ValueError("Parser disposed"))
        self._emit_log(
            ParseEvent(
                type=ParseEventType.DISPOSED,
                property_path="",
                message="Parser disposed",
            )
        )

    def _ensure_not_disposed(self) -> None:
        if self._disposed:
            raise RuntimeError("Parser has been disposed")

    def _add_property_chunk(self, property_path: str, chunk: Any) -> None:
        controller = self._property_controllers.get(property_path)
        if not controller:
            return

        if isinstance(controller, StringPropertyStreamController):
            controller.add_chunk(str(chunk))
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.STRING_CHUNK,
                    property_path=property_path,
                    message="String chunk received",
                    data=chunk,
                )
            )
        elif isinstance(controller, MapPropertyStreamController):
            controller.add_snapshot(chunk)
        elif isinstance(controller, ListPropertyStreamController):
            controller.add_snapshot(chunk)

    def _get_controller_for_path(
        self, property_path: str
    ) -> Optional[PropertyStreamController[Any]]:
        return self._property_controllers.get(property_path)

    def _get_property_stream(self, property_path: str, stream_type: str) -> PropertyStream[Any]:
        existing = self._property_controllers.get(property_path)
        if existing:
            if self._is_compatible(existing, stream_type):
                return existing.property_stream
            requested = existing.__class__.__name__.replace("PropertyStreamController", "")
            error_message = (
                f"Type mismatch at path \"{property_path}\": requested {requested} "
                f"but found {stream_type} in JSON"
            )
            if self._saw_potential_thinking_tags and not self.skip_thoughts:
                error_message += (
                    "\n\nHint: The input may contain thinking tags. "
                    "Try setting skip_thoughts=True."
                )
            error = ValueError(error_message)
            existing.complete_error(error)
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.ERROR,
                    property_path=property_path,
                    message="Type mismatch",
                    data=error,
                )
            )
            raise error

        if stream_type == "string":
            return self.get_string_property(property_path)
        if stream_type == "number":
            return self.get_number_property(property_path)
        if stream_type == "boolean":
            return self.get_boolean_property(property_path)
        if stream_type == "null":
            return self.get_null_property(property_path)
        if stream_type == "object":
            return self.get_map_property(property_path)
        if stream_type == "array":
            return self.get_list_property(property_path)
        raise ValueError(f"Unknown stream type: {stream_type}")

    def _complete_property(self, property_path: str, value: Any) -> None:
        controller = self._property_controllers.get(property_path)
        if controller:
            controller.complete(value)
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.PROPERTY_COMPLETE,
                    property_path=property_path,
                    message="Property completed",
                )
            )
        if property_path == "":
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.ROOT_COMPLETE,
                    property_path="",
                    message="Root JSON complete",
                )
            )

    def _is_compatible(self, controller: PropertyStreamController[Any], stream_type: str) -> bool:
        return (
            (stream_type == "string" and isinstance(controller, StringPropertyStreamController))
            or (stream_type == "number" and isinstance(controller, NumberPropertyStreamController))
            or (stream_type == "boolean" and isinstance(controller, BooleanPropertyStreamController))
            or (stream_type == "null" and isinstance(controller, NullPropertyStreamController))
            or (stream_type == "object" and isinstance(controller, MapPropertyStreamController))
            or (stream_type == "array" and isinstance(controller, ListPropertyStreamController))
        )

    def _parse_chunk(self, chunk: str) -> None:
        if self._disposed:
            return
        if self.close_on_root_complete and self._root_delegate and self._root_delegate.done:
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.YAP_FILTERED,
                    property_path="",
                    message="Yap filter triggered - ignoring text after root JSON",
                )
            )
            self._abort_stream = True
            return

        try:
            for character in chunk:
                if (
                    self.close_on_root_complete
                    and self._root_delegate
                    and self._root_delegate.done
                ):
                    self._emit_log(
                        ParseEvent(
                            type=ParseEventType.YAP_FILTERED,
                            property_path="",
                            message="Yap filter triggered - ignoring text after root JSON",
                        )
                    )
                    self._abort_stream = True
                    return

                if self.skip_thoughts:
                    processed = self._process_character_for_thinking_tags(character)
                    if processed is None:
                        continue
                    character = processed
                else:
                    self._detect_potential_thinking_tags(character)

                if self._root_delegate:
                    self._root_delegate.add_character(character)
                    continue

                if character.isspace():
                    continue

                if character == "{":
                    self._emit_log(
                        ParseEvent(
                            type=ParseEventType.ROOT_START,
                            property_path="",
                            message="Started parsing root object",
                        )
                    )
                    self._root_delegate = ObjectPropertyDelegate(
                        "", self._controller
                    )
                    self._root_delegate.add_character(character)
                    continue

                if character == "[":
                    self._emit_log(
                        ParseEvent(
                            type=ParseEventType.ROOT_START,
                            property_path="",
                            message="Started parsing root array",
                        )
                    )
                    self._root_delegate = ArrayPropertyDelegate(
                        "", self._controller
                    )
                    self._root_delegate.add_character(character)
                    continue

            if self._root_delegate:
                self._root_delegate.on_chunk_end()

            if self.close_on_root_complete and self._root_delegate and self._root_delegate.done:
                self._emit_log(
                    ParseEvent(
                        type=ParseEventType.YAP_FILTERED,
                        property_path="",
                        message="Yap filter triggered - root JSON complete",
                    )
                )
                self._abort_stream = True
        except Exception as exc:  # pragma: no cover - defensive
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.ERROR,
                    property_path="",
                    message=f"Parsing error: {exc}",
                    data=exc,
                )
            )

    def _handle_stream_end(self) -> None:
        for controller in self._property_controllers.values():
            if not controller.is_closed:
                controller.complete_error(
                    RuntimeError("Stream ended before property completed")
                )

    def _emit_log(self, event: ParseEvent) -> None:
        if self._on_log:
            self._on_log(event)

    def _process_character_for_thinking_tags(self, character: str) -> Optional[str]:
        start_tag, end_tag = self.thinking_tags

        if self._inside_thinking_tags:
            self._tag_buffer += character
            if self._tag_buffer.endswith(end_tag):
                self._inside_thinking_tags = False
                self._tag_buffer = ""
                self._emit_log(
                    ParseEvent(
                        type=ParseEventType.THINKING_TAG_END,
                        property_path="",
                        message="Exited thinking tags",
                    )
                )
                return None

            if len(self._tag_buffer) > len(end_tag) * 2:
                self._tag_buffer = self._tag_buffer[-len(end_tag) :]
            return None

        self._tag_buffer += character
        if self._tag_buffer.endswith(start_tag):
            self._inside_thinking_tags = True
            self._tag_buffer = ""
            self._emit_log(
                ParseEvent(
                    type=ParseEventType.THINKING_TAG_START,
                    property_path="",
                    message="Entered thinking tags",
                )
            )
            return None

        if len(self._tag_buffer) > len(start_tag):
            self._tag_buffer = self._tag_buffer[-len(start_tag) :]
            return character

        if start_tag.startswith(self._tag_buffer):
            return None

        self._tag_buffer = ""
        return character

    def _detect_potential_thinking_tags(self, character: str) -> None:
        if self._root_delegate:
            return

        start_tag, _ = self.thinking_tags
        self._potential_tag_buffer += character
        if self._potential_tag_buffer.endswith(start_tag):
            self._saw_potential_thinking_tags = True
            self._potential_tag_buffer = ""
            return

        if len(self._potential_tag_buffer) > len(start_tag):
            self._potential_tag_buffer = self._potential_tag_buffer[-len(start_tag) :]
